README.md
# jose

[![Build](https://travis-ci.org/autoit4you/node-jose.svg?branch=master)](https://travis-ci.org/autoit4you/node-jose)
[![Coverage](https://coveralls.io/repos/autoit4you/node-jose/badge.svg?branch=master&service=github)](https://coveralls.io/github/autoit4you/node-jose?branch=master)

# API
## JWA
The implementation of the 
[JSON Web Algorithms](https://tools.ietf.org/html/rfc7518) for use
with [JSON Web Signatures](https://tools.ietf.org/html/rfc7515) or 
[JSON Web Encryption](https://tools.ietf.org/html/rfc7516).

The following digital signature/MAC algorithms are supported:

"alg" Header Parameter | Digital Signature or MAC Algorithm
-----------------|-----------------------------------
HS256 | HMAC using SHA-256 hash algorithm
HS384 | HMAC using SHA-384 hash algorithm
HS512 | HMAC using SHA-512 hash algorithm
RS256 | RSASSA-PKCS1-v1_5 using SHA-256 hash algorithm
RS384 | RSASSA-PKCS1-v1_5 using SHA-384 hash algorithm
RS512 | RSASSA-PKCS1-v1_5 using SHA-512 hash algorithm
ES256 | ECDSA using P-256 curve and SHA-256 hash algorithm
ES384 | ECDSA using P-384 curve and SHA-384 hash algorithm
ES512 | ECDSA using P-521 curve and SHA-512 hash algorithm
none | No integrity protection

The following key encryption algorithms are supported:

"alg" Header Parameter | Key Encryption Algorithm | Additional paramters needed
-----------------------|------------------------- | ---------------------------
RSA1_5 | Key wrapping using RSAES-PKCS1-v1_5 | none
dir | Direct use of key as CEK | none
A128GCMKW | Key wrapping using AES-GCM with 128-bit key | iv, tag for unwrapping
A192GCMKW | Key wrapping using AES-GCM with 192-bit key | iv, tag for unwrapping
A256GCMKW | Key wrapping using AES-GCM with 256-bit key | iv, tag for unwrapping

The following content encryption algorithms are supported:

"enc" Header Parameter | Content Encryption Algorithm
-----------------|-----------------------------------------
A128CBC-HS256 | AES CBC using 128-bit key with SHA-256 as HMAC
A192CBC-HS384 | AES CBC using 192-bit key with SHA-384 as HMAC
A256CBC-HS512 | AES CBC using 256-bit key with SHA-512 as HMAC
A128GCM | AES GCM using 128-bit key
A192GCM | AES GCM using 192-bit key
A256GCM | AES GCM using 256-bit key

### jwa(algorithm)
Creates a new `jwa` object with `sign` and `verify` or `encrypt` and `decrypt` methods, depending 
on the specified algorithm. Valid values for `algorithm` can be found 
in the above table and are case-sensitive. Passing an invalid algorithm 
will throw a `TypeError`.

#### Example:
```js
var jose = require('jose');
var hs256 = jose.jwa('HS256');
```

### jwa.sign(input, secretOrPrivateKey)
*This method is only available to algorithms for digital signatures and MACs.*

Sign `input` with either a secret using a HMAC algorithm or
a *private* key using a digital signature algorithm.

`input` must be a buffer.

When using a HMAC algorithm `secretOrPrivateKey` should be either a
string or a buffer. When using a digital signature algorithm, the
value should be a PEM encoded *private* key.

Returns the signature as buffer.

#### Example:
```js
var jose = require('jose');
var hs256 = new jose.jwa('HS256');
var signature = hs256.sign(input, 'secret');
console.log(signature); // Prints Buffer <8C, ...
```

### jwa.verify(input, signature, secretOrPublicKey)
*This method is only available to algorithms for digital signatures and MACs.*

Verify that `signature` is a valid signature for `input` using 
a secret for HMAC algorithms or a *public* key for 
digital signature algorithms.

`input` must be a buffer.

When using a HMAC algorithm `secretOrPublicKey` should be either a
string or a buffer. When using a digital signature algorithm, the
value should be a PEM encoded *public* key.

Returns a boolean indicating whether or not the signature is valid.

#### Example:
```js
var jose = require('jose');
var hs256 = new jose.jwa('HS256');
var isValid = hs256.verify(input, signature, 'secret');
console.log(isValid); // Prints 'true'
```

### jwa.encrypt(plaintext, aad, iv, key)
*This method is only available to algorithms for content encryption.*

Encrypt `plaintext` and compute an authentication tag using `aad`,
`iv` and `key`.

All parameters must be `Buffer`s except `plaintext` which can
also be a `string`.

Returns an object containing `cipher` with the ciphertext and `tag`
with the authentication tag, both being a `Buffer`.

#### Example:
```js
var jose = require('jose');
var aes = new jose.jwa('A256GCM');
var enc = aes.encrypt('secret', aad, iv, key);
console.log(enc.cipher); // Prints Buffer <5F, ...
console.log(enc.tag); // Prints Buffer <9E, ...
```

### jwa.decrypt(ciphertext, tag, aad, iv, key)
*This method is only available to algorithms for content encryption.*

Authenticate and decrypt `ciphertext` using `tag`, `aad`, `iv` and `key`.

All parameters must be `Buffer`s.

Returns a `Buffer` containing the decrypted plaintext. Throws an
Error if the `ciphertext` can't be authenticated.

#### Example:
```js
var jose = require('jose');
var aes = new jose.jwa('A256GCM');
var plain = aes.encrypt(ciphertext, tag, aad, iv, key);
console.log(plain); // Prints Buffer <F9, ...
```

### jwa.wrapKey(keyToWrap, key[, options])
*This method is only available to algorithms for key encryption.*

Wraps the `keyToWrap` with a key suitable to the used algorithm,
as defined by RFC7518. `options` may have to be provided, 
depending on the algorithm used for wrapping.

`keyToWrap` must be a `Buffer`.
`key` must be of an type that the used algorithm accepts.
`options`, if provided, must be an object.

Returns an object containing `keydata`, which is a `Buffer`, 
and possibly further arguments that are needed to unwrap the key again 
(see supported algorithms table).

#### Example:
```js
var jose = require('jose');
var rsa = new jose.jwa('RSA1_5');
var wrapped = rsa.wrapKey(keyToWrap, key);
console.log(wrapped.keydata); // Prints Buffer <58, ...
```

### jwa.unwrapKey(keyToUnwrap, key[, options])
*This method is only available to algorithms for key encryption.*

Unwraps the `keyToUnwrap` with a key suitable to the used algorithm,
as defined by RFC7518. `options` may have to be provided, 
depending on the algorithm used for unwrapping.

`keyToUnwrap` must be a `Buffer`.
`key` must be of an type that the used algorithm accepts.
`options`, if provided, must be an object.

Returns a `Buffer`, which is the unwrapped key.

#### Example:
```js
var jose = require('jose');
var rsa = new jose.jwa('RSA1_5');
var unwrapped = rsa.unwrapKey(keytoUnwrap, key);
console.log(unwrapped); // Prints Buffer <A7, ...
```


CHANGELOG.md
# jose ChangeLog

## 2015-12-06, Version 0.3.2

### Notable Changes

None

### Known Issues

None

### Commits

* [`c445eb3`] Release 0.3.1
* [`dbc6db2`] Amended documentation
* [`1466535`] Release 0.3.0

## 2015-12-01, Version 0.3.0

### Notable Changes

* **jwa**:
  - Added support for the Key Encryption Algorithms RSA1_5, dir, A128KWGCM, 
      A192KWGCM and A256KWGCM, as specified in the JWA RFC
  - The signature algorithms now return and expect buffers instead of 
      Base64Url-encoded strings

### Known Issues

* README.md not fully up-to-date

### Commits

* [`dd9c2db`] Added documentation for jwa.unwrapKey
* [`678e50f`] Added documentation for jwa.wrapKey
* [`3b73781`] Added support for A*KWGCM and RSA1_5 algorithms
* [`d26db82`] Direct Encryption can't have a wrong key
* [`bf9ecba`] test the examples from the README
* [`085c87a`] Start work on Key Encryption Algorithms
* [`eb0aa3e`] Recommend usage of new
* [`7084b3a`] Fix README for signatures
* [`8836326`] Signature methods now expect and return buffers

## 2015-09-22, Version 0.2.0

### Notable Changes

* **jwa**:
  - Added support for all Content Encryption Algorithms specified in the JWA RFC

### Known Issues

None

### Commits

* [`945d586`] Preparing 0.2.0 release
* [`1d4f28f`] Added changelog
* [`48497ef`] Corrected error in README
* [`6e26988`] Added decrypt section to README
* [`60c6230`] Added encrypt section to README
* [`671368c`] Implemented decrypting using AES-GCM
* [`22545a6`] Implemented decrypting AES-CBC
* [`819af8a`] Drop support for 0.10
* [`2ad8eae`] Adjust build versions
* [`d0f6653`] Fix build failure on 0.10
* [`a66c086`] Fixed AES-GCM encryption
* [`f6e05c0`] Started work on Content Encryption Algorithms
* [`8300488`] Added bn.js to dependencies
* [`052d60d`] Removed unneccessary title

## 2015-09-01, Version 0.1.0

### Notable Changes

* First release


### Known Issues

None

### Commits

* [`47e2a97`] Corrected error in README
* [`b05374c`] Completed readme
* [`12d6675`] Added note about stringification of input
* [`b2e258f`] Now normalizing the input before processing it
* [`e3c05fb`] Prepared library for first release
* [`a08cdcc`] Added jwa.sign documentation
* [`53ed188`] Added Build and Coverage badges
* [`d798ac7`] Worked on the README
* [`31d2806`] Completed ES* signature verification
* [`b48772c`] Implemented signing using the ES* algorithms
* [`9859a61`] Started work on ES signatures
* [`0058847`] Implemented RSASSA-PKCS1-v1_5 signatures with tests
* [`2f2f49b`] Added toBase64 method
* [`34a2929`] Changed Travis build cmds
* [`287dce8`] Taking a new approach on this. Now supports JWA algs none and HMAC.
* [`ed6a3f6`] Setup Travis CI
* [`246c1d8`] Corrected version
* [`c5b3e73`] Added watcher tests
* [`fc3f6fe`] Finished JWK, for now
* [`ffed23d`] Abandoned the key-reader will instead use OpenSSL to do this, later...
* [`c5c75e9`] Started work on BER parser
* [`44d4d53`] Finalized Hmac Verifier
* [`9843538`] Started work on JWKs
* [`ee707ea`] Started work on Hmac verifier. Added base64url as dependency.
* [`704becb`] Now using factories
* [`25d6179`] Added Readme
* [`fa00c25`] Initial commit


lib/jwa/signer.js
"use strict";
var crypto = require("crypto");
var base64url = require("../base64url");
var bufferEqual = require("../bufferEqual");
var BigNumber = require("bn.js");
var EcdsaSignature = require("asn1.js").define("Ecdsa-Sig-Value", function() {
	this.seq().obj(
		this.key("r").int(),
		this.key("s").int()
	);
});

var ecdsaKeySize = {
	"256": 256,
	"384": 384,
	"512": 521
}

var signer = {
	"none": createNoneSigner,
	"HS": createHsSigner,
	"RS": createRsSigner,
	"ES": createEsSigner
}

var verifier = {
	"none": createNoneVerifier,
	"HS": createHsVerifier,
	"RS": createRsVerifier,
	"ES": createEsVerifier
}

function verifyBuffer(buf, param) {
	if(!Buffer.isBuffer(buf)) {
		throw new Error(param + " must be a buffer");
	}
}

function createNoneSigner() {
	return function sign(input) {
		verifyBuffer(input, "input");
		return new Buffer(0);
	}
}

function createNoneVerifier() {
	return function verify(input, signature) {
		verifyBuffer(input, "input");
		verifyBuffer(signature, "signature");
		return signature.length === 0;
	}
}

function createHsSigner(bits) {
	return function sign(input, secret) {
		verifyBuffer(input, "input");
		var hmac = crypto.createHmac("SHA" + bits, secret);
		hmac.update(input);
		return hmac.digest();
	}
}

function createHsVerifier(bits) {
	var signer = createHsSigner(bits);
	return function verify(input, signature, secret) {
		verifyBuffer(input, "input");
		verifyBuffer(signature, "signature");
		var sig2 = signer(input, secret);
		return bufferEqual(signature, sig2);
	}
}

function createRsSigner(bits) {
	return function sign(input, key) {
		verifyBuffer(input, "input");
		var signer = crypto.createSign("RSA-SHA"+bits);
		signer.update(input);
		return signer.sign(key);
	}
}

function createRsVerifier(bits) {
	return function verify(input, signature, key) {
		verifyBuffer(input, "input");
		verifyBuffer(signature, "signature");

		var verifier = crypto.createVerify("RSA-SHA"+bits);
		verifier.update(input);
		return verifier.verify(key, signature);
	}
}

function createEsSigner(bits) {
	return function sign(input, key) {
		verifyBuffer(input);

		var signer = crypto.createSign("RSA-SHA"+bits); // openssl only signs with ec with sha2 when specifying rsa and using an ec key
		signer.update(input);
		var sig = signer.sign(key);
		var derSig = EcdsaSignature.decode(sig, "der");

		var bytesLength = (ecdsaKeySize[bits]/8 | 0) + (ecdsaKeySize[bits]%8 !== 0 ? 1 : 0);
		derSig.r = new Buffer(derSig.r.toString("hex", bytesLength), "hex");
		derSig.s = new Buffer(derSig.s.toString("hex", bytesLength), "hex");

		return Buffer.concat([derSig.r, derSig.s], derSig.r.length + derSig.s.length);
	}
}

function createEsVerifier(bits) {
	return function verify(input, signature, key) {
		verifyBuffer(input, "input");
		verifyBuffer(signature, "signature");

		var bytesLength = (ecdsaKeySize[bits]/8 | 0) + (ecdsaKeySize[bits]%8 !== 0 ? 1 : 0);
		if(signature.length !== bytesLength*2) {
			return false;
		}

		var r = new BigNumber(signature.slice(0, bytesLength), 10, "be").iabs();
		var s = new BigNumber(signature.slice(bytesLength), 10, "be").iabs();


		var derSig = EcdsaSignature.encode({
			"r": r,
			"s": s
		}, "der");

		var verifier = crypto.createVerify("RSA-SHA"+bits); // See above for explanation
		verifier.update(input);
		return verifier.verify(key, derSig, "base64");
	}
}

module.exports = function(alg, bits) {
	return {
		sign: signer[alg](bits),
		verify: verifier[alg](bits)
	}
}